### PR TITLE
HADOOP-18066: AbstractJavaKeyStoreProvider: need a way to read credential store password from Configuration

### PR DESCRIPTION
**Pull request by abstractdog**
Change-Id: I272dd387ecb52eccd8035661cfe35edcdb29840c

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [x] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 59s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  35m 20s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  24m 12s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  20m 44s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   1m  1s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 35s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  8s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 28s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 10s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 59s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  23m 29s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |  23m 29s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 42s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  20m 42s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  1s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 36s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  6s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 40s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 42s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m 19s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  17m 18s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  |  The patch does not generate ASF License warnings.  |
|  |   | 210m 41s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ipc.TestIPC |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3865 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux e4868e0790f8 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / e7da7a861ab061fa732df33df463dddfeaf5ee30 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/1/testReport/ |
| Max. process+thread count | 1252 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by abstractdog** [comment]
TestIPC failure doesn't seem to be related

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java [hunk 1]
```diff
@@ -2485,7 +2485,7 @@ private CredentialEntry getCredentialEntry(CredentialProvider provider,
    * @param name
    * @return clear text password or null
    */
-  protected char[] getPasswordFromConfig(String name) {
+  public char[] getPasswordFromConfig(String name) {
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java:2493]: is there any risk of recursive store init going on?

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/CommonConfigurationKeysPublic.java [hunk 2]
```diff
@@ -988,6 +988,14 @@
   public static final String  HADOOP_SECURITY_CREDENTIAL_PASSWORD_FILE_KEY =
       "hadoop.security.credstore.java-keystore-provider.password-file";
 
+  /**
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/CommonConfigurationKeysPublic.java:991]: this will need docs with the other credential options

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/AbstractJavaKeyStoreProvider.java [hunk 3]
```diff
@@ -308,15 +311,24 @@ public void flush() throws IOException {
   /**
    * Open up and initialize the keyStore.
    *
+   * Password evaluation precedence order:
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/AbstractJavaKeyStoreProvider.java:314]: should be an html numbered list

**Review by steveloughran — CHANGES_REQUESTED** [review]
some minor comments.

1. this should be testable
2. we should mark the exposed method as `@Private @Unstable` so we don't have to support/maintain it forever. the javadocs should say that it is only for use in the hadoop security code

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java (hunk 1 above)
Comment (abstractdog) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java:2493]: when I used conf.getPassword(), it went to an infinite recursion (init keystore -> go to credential provider -> init keystore ...), that's how I ended up with getPasswordFromConfig

**Comment by abstractdog** [comment]
thanks a lot @steveloughran for the comments! I've added a new commit addressing those

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  18m 28s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  35m 48s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  24m 26s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  20m 40s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   1m  4s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 37s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 35s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 29s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 12s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  1s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  23m 30s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |  23m 30s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 41s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  20m 41s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m  0s | [/results-checkstyle-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/2/artifact/out/results-checkstyle-hadoop-common-project_hadoop-common.txt) |  hadoop-common-project/hadoop-common: The patch generated 2 new + 322 unchanged - 0 fixed = 324 total (was 322)  |
| +1 :green_heart: |  mvnsite  |   1m 34s |  |  the patch passed  |
| +1 :green_heart: |  xml  |   0m  1s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   1m  4s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 39s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 44s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m 53s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 11s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  |  The patch does not generate ASF License warnings.  |
|  |   | 228m  3s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3865 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell xml markdownlint |
| uname | Linux b1b4a4ba3ffc 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / c207f7719c436026e799e9c635d21934fbf00213 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/2/testReport/ |
| Max. process+thread count | 1363 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  12m 18s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m 48s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  22m 22s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  19m 28s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   1m  9s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 39s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 13s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 47s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 28s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  22m  1s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 58s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 41s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |  21m 41s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 28s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  19m 28s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m  8s | [/results-checkstyle-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/3/artifact/out/results-checkstyle-hadoop-common-project_hadoop-common.txt) |  hadoop-common-project/hadoop-common: The patch generated 2 new + 322 unchanged - 0 fixed = 324 total (was 322)  |
| +1 :green_heart: |  mvnsite  |   1m 38s |  |  the patch passed  |
| +1 :green_heart: |  xml  |   0m  1s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   1m 42s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   2m 35s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  22m  6s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  17m 22s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/3/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  0s |  |  The patch does not generate ASF License warnings.  |
|  |   | 208m 21s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ipc.TestIPC |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3865 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell xml markdownlint |
| uname | Linux 5b8b750d26fd 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / e7f7d5a65b42f41e28f4e7eb1af6d72957d82ca9 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/3/testReport/ |
| Max. process+thread count | 3152 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3865/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java (hunk 1 above)
Comment (steveloughran) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java:2493]: aah. filesytems are meant to undeclare themselves as a source of credentials to avoid this...was this secrets for an fs/object store you were trying to retrieve?

**Comment by lmccay** [comment]
The whole purpose for the credential provider API and framework here is to remove secrets/passwords from the configuration. Why are we looking to add the password to the credential store to the configuration here? It seems that it completely defeats the purpose of the feature. I would rather see the default credential store password be used and solely rely on the file permissions of the credential store to protect who has access. What am I missing?

**Comment by abstractdog** [comment]
@lmccay , @steveloughran : thanks for your comments
TLDR: finally I made this work on hive side (HIVE-25829) without this workaround, so I'll close this PR and jira

I understand that password in conf object is against design, because, for instance, it clearly states that a single password should be used for the providers
when I started to work on this workaround, I was confused about HIVE-14822, which looks like messing up the hadoop design by introducing more than 1 credential stores with different passwords (1 for hiveserver2, 1 to distribute on cluster for mr/spark/tez jobs), however it's not really the case
after fixing my code, I properly use HADOOP_CREDSTORE_PASSWORD in hiveserver2, and propagate the other password (together with the distributed provider path) to the jobs

**Comment by lmccay** [comment]
@abstractdog - very good - thanks for the update!
